### PR TITLE
refactor: block-based code-splitting

### PR DESCRIPTION
**Pull request by ahabhgk**
## Summary

- block based code-splitting, trying my best to align exist code-splitting logic with webpack 🥲
- block_promise for block which contains multiple dependencies
- refactor ContextModule codegen based on new block_promise
- MergeDuplicateChunksPlugins for fixing a regression due to this refactor, before it's coupled into code-splitting, now de-coupled from it and aligned with webpack
- fix unstable chunkIds for those generate same chunkId chunks, should use `compare_chunks_natural` instead of compare chunk.name 

## Require Documentation?

- [x] No
- [ ] Yes, the corresponding rspack-website PR is \_\_

### CODE REVIEW

File: crates/rspack/tests/samples/remove-parent-modules/cycle-dynamic-entry/snapshot/output.snap [hunk 1]
```diff
@@ -19,15 +19,45 @@ console.log('shared');
 }]);
 ```
 
-```js title=dynamic-2_js.js
-(self['webpackChunkwebpack'] = self['webpackChunkwebpack'] || []).push([["dynamic-2_js"], {
+```js title=dynamic-1_js~_1d741.js
```
Comment (ahabhgk) [inline crates/rspack/tests/samples/remove-parent-modules/cycle-dynamic-entry/snapshot/output.snap:22]: Aligned with webpack

File: crates/rspack/tests/samples/remove-parent-modules/intersection/snapshot/output.snap [hunk 2]
```diff
@@ -17,16 +17,36 @@ __webpack_require__.r(__webpack_exports__);
...
-console.log('i-1');
-}),
 "./i-2.js": (function (__unused_webpack_module, exports, __webpack_require__) {
 console.log('i-2');
 }),
 
 }]);
 ```
 
+```js title=a_js~_9b1b1.js
```
Comment (ahabhgk) [inline crates/rspack/tests/samples/remove-parent-modules/intersection/snapshot/output.snap:27]: Aligned with webpack

**Comment by ahabhgk** [comment]
!eco-ci

**Comment by ahabhgk** [comment]
modernjs failed because `optimization.chunkIds` set to deterministic

**Review by bvanjoi — APPROVED** [review]

**Comment by hardfist** [comment]
this is a big refactor, needs more people review or setup a CR meeting

**Comment by jerrykingxyz** [comment]
!bench

**Pull request merged**